docs/current_docs/manuals/developer/functions.mdx

```diff
@@ -338,19 +338,186 @@ The result will look like this:
 Hello, world
 ```
 
+### Directories and files
+
+It is possible to automatically load a filesystem path as a `Directory` or `File` object in a Dagger Function, by passing it as a default context to the corresponding argument. The `Directory` or `File` loaded in this manner is not merely a string, but it is the actual filesystem state of the specified directory or file, managed by the Dagger Engine and handled in code just like any another variable.
+
+:::important
+Default contexts are only available for arguments of type `Directory` and `File`. They are commonly used to load constant filesystem locations, such as an application's source code directory.
+:::
+
+The default context is set by adding a `defaultPath` pragma (Go), `DefaultPath` annotation (Python), or `defaultPath` decorator (TypeScript) on the corresponding Dagger Function argument.
+
+The default context is resolved depending on various factors:
+- If an absolute path is specified:
+   - in a Git repository (defined by the presence of a `.git` sub-directory), the default context is the root of the Git repository.
+   - in a non-repository location (defined by the absence of a `.git` sub-directory), the default context is the directory containing a `dagger.json` file.
+- If a relative path is specified
+  - the default context is always the directory containing a `dagger.json` file.
+
+Here is an example of a Dagger Function with a default context for its directory argument:
+
+<Tabs groupId="language">
+<TabItem value="Go">
+```go file=./snippets/functions/arguments-contextual-dir/go/main.go
+```
+</TabItem>
+<TabItem value="Python">
+```python file=./snippets/functions/arguments-contextual-dir/python/main.py
+```
+</TabItem>
+<TabItem value="TypeScript">
+```typescript file=./snippets/functions/arguments-contextual-dir/typescript/index.ts
+```
+</TabItem>
+</Tabs>
+
+Assuming the directory structure below, the following table describes how the argument is resolved:
+
+```shell
+.
+├── README.md
+├── my-module
+│   ├── sdk
+│   ├── dagger.json
+│   ├── ...
+├── .git
+│   └── ...
+├── index.html
+├── public
+│   └── ...
+├── src
+│   └── ...
+```
+
+| Location | Default context ('defaultPath`) | Dagger Function invocation | `source` resolves to |
+| --- | --- | --- | --- |
+| Git repository | `/` | `dagger call -m my-module read-dir` | `/` |
+| Git repository | `/src` | `dagger call -m my-module read-dir` | `/src` |
+| Git repository | `.` | `dagger call -m my-module read-dir` | `/my-module` |
+| Git repository | `..` | `dagger call -m my-module read-dir` | `/` |
+| Non-Git directory | `/` | `dagger call -m my-module read-dir` | `/` |
+| Non-Git directory | `/src` | `dagger call -m my-module read-dir` | `/src` |
+| Non-Git directory | `.` | `dagger call -m my-module read-dir` | `/my-module` |
+| Non-Git directory | `..` | `dagger call -m my-module read-dir` | `/` |
```

Comment:
See https://github.com/dagger/dagger/pull/8339#issuecomment-2343853278 for context 
```suggestion
| Git repository | `/` | `dagger call -m my-module read-dir` | `/` |
| Git repository | `/src` | `dagger call -m my-module read-dir` | `/src` |
| Git repository | `.` | `dagger call -m my-module read-dir` | `/my-module` |
| Git repository | `..` | `dagger call -m my-module read-dir` | `/` |
| Non-Git directory | `/` | `dagger call -m my-module read-dir` | `/` |
| Non-Git directory | `/src` | `dagger call -m my-module read-dir` | ❌ will fail because /src is outside of the root dir context |
| Non-Git directory | `.` | `dagger call -m my-module read-dir` | `/my-module` |
| Non-Git directory | `..` | `dagger call -m my-module read-dir` | ❌ will fail because .. escape from the root dir context |
```